cvmfs/export_plugin/fs_traversal.cc

```diff
@@ -0,0 +1,793 @@
+ * 
+ * This method will return true exactly once for each inode and can therefore
+ * be used as an atomic locking procedure for file writing decisions.
+ */
+bool should_write_anyway(struct fs_traversal *dest,
+  struct cvmfs_attr *src_st,
+  struct cvmfs_attr *dest_st) {
+  if (!dest->is_hash_consistent(dest->context_, dest_st)
+    && fsck_lock->AddLock(dest_st->st_ino)) {
+    return true;
+  }
+  return false;
+}
+
+bool handle_file(
+  struct fs_traversal *src,
+  struct cvmfs_attr *src_st,
+  struct fs_traversal *dest,
+  struct cvmfs_attr *dest_st,
+  const char *entry,
+  perf::Statistics *pstats
+) {
+  bool result = true;
+  // They don't point to the same data, link new data
+  char *dest_data = dest->get_identifier(dest->context_, src_st);
+
+  // Touch is atomic, if it fails something else will write file
+  if (!dest->touch(dest->context_, src_st)
+    || (dest_st->cvm_checksum != NULL
+      && should_write_anyway(dest, src_st, dest_st)) ) {
+    char *src_ident = src->get_identifier(src->context_, src_st);
+    if (num_parallel_) {
+      FileCopy next_copy(src_ident, strdup(dest_data));
+
+      WritePipe(pipe_chunks[1], &next_copy, sizeof(next_copy));
+      atomic_inc64(&copy_queue);
+    } else {
+      if (!copyFile(src, src_ident, dest, dest_data, pstats)) {
+        LogCvmfs(kLogCvmfs, kLogStderr,
+          "Failed to copy %s->%s : %d : %s",
+          entry, dest_data, errno, strerror(errno));
+        errno = 0;
+        result = false;
+      }
+      free(src_ident);
+      pstats->Lookup(SHRINKWRAP_STAT_FILE_COUNT)->Inc();
+    }
+  } else {
+    pstats->Lookup(SHRINKWRAP_STAT_DEDUPED_FILES)->Inc();
+    pstats->Lookup(SHRINKWRAP_STAT_DEDUPED_BYTES)->Xadd(src_st->st_size);
+  }
+
+  // Should probably happen in the copy function as it is parallel
+  // Also this needs to be separate from copyFile, the target file
+  // could already exist and the link needs to be created anyway.
+  if (result && dest->do_link(dest->context_, entry, dest_data)) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+      "Failed to link %s->%s : %d : %s",
+      entry, dest_data, errno, strerror(errno));
+    errno = 0;
+    result = false;
+  }
+  free(dest_data);
+  return result;
+}
+
+bool handle_dir(
+  struct fs_traversal *src,
+  struct cvmfs_attr *src_st,
+  struct fs_traversal *dest,
+  struct cvmfs_attr *dest_st,
+  const char *entry
+) {
+  if (dest->do_mkdir(dest->context_, entry, src_st)) {
+    if (errno == EEXIST) {
+      errno = 0;
+      if (dest->set_meta(dest->context_, entry, src_st)) {
+        LogCvmfs(kLogCvmfs, kLogStderr,
+          "Traversal failed to set_meta %s", entry);
+        return false;
+      }
+    } else {
+      LogCvmfs(kLogCvmfs, kLogStderr,
+        "Traversal failed to mkdir %s : %d : %s",
+        entry, errno, strerror(errno));
+      return false;
+    }
+  }
+  return true;
+}
+
+void add_dir_for_sync(const char *dir, bool recursive) {
```

Comment:
This function may not be necessary, but I left it here in case we ever change the structure used to store the directories.